java_src/foam/dao/index/TreeIndex.java

```diff
@@ -0,0 +1,152 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package foam.dao.index;
+
+import foam.core.FObject;
+import foam.core.PropertyInfo;
+import foam.dao.Sink;
+import foam.mlang.predicate.And;
+import foam.mlang.predicate.Binary;
+import foam.mlang.predicate.Eq;
+import foam.mlang.predicate.False;
+import foam.mlang.predicate.Gt;
+import foam.mlang.predicate.Gte;
+import foam.mlang.predicate.Lt;
+import foam.mlang.predicate.Lte;
+import foam.mlang.predicate.Predicate;
+import foam.mlang.sink.Count;
+import java.util.Comparator;
+
+public class TreeIndex implements Index {
+  protected Index tail;
+  protected PropertyInfo prop;
+  protected long selectCount;
+  protected TreeNode tree;
+  
+  public TreeIndex(PropertyInfo prop) {
+    this.prop = prop;
+    this.selectCount = 0;
+    this.tail = ValueIndex.instance();
+    this.tree = TreeNode.getNullNode();
+  }
+  
+  public TreeIndex(PropertyInfo prop, Index tail) {
+    this.prop = prop;
+    this.selectCount = 0;
+    this.tail = tail;
+  }
+  
+  public void bulkLoad(FObject... a) {
+    if ( this.tail instanceof ValueIndex ) {
+      this.tree = (TreeNode)this.tree.bulkLoad(this.prop, 0, a.length-1, a);
+    }
+    for ( int i = 0 ; i < a.length ; i++ ) {
+      this.tree = (TreeNode) this.put(this.tree, a[i]);
+    }
+  }
+  // predicate from plan, prop from index itself, Mlang Expression
```

Comment:
Better docs and maybe better names needed here. Shouldn't one of the arguments be a Class?